### ISSUE TITLE
Is the restored state being merged with the default?

### ISSUE THREAD

**Issue by sp00x** [championswimmer/vuex-persist#106]
This might be documented somewhere, but I couldn't find it now, and it was a bit of a "huh?" moment here:

I am persisting some settings related to UI layout, where the user can chose which things appear in which column, e.g. I had something like

```
state: {
  leftSide: [ "a", "b" ], // default layout
  rightSide: [ "c" ] // default layout
}
```

However, it seems every time the state is being restored it is not *overwritten* but *merged*, so that if the user had changed stuff, e.g. moved "a" to rightSide it ends up popping back into leftSide (along with being in rightSide) also upon restore.

Is there any settings/options to have vuex-persist *overwrite* instead of merge on restore? Right now I have to have a default state with blank arrays, and an action to check if they're blank and populate with defaults when I set up the store..

**Comment by tmladek** [comment]
I am having the exact same issue and it's been driving me crazy.

I would consider it a bug, as there's no indication this is the behaviour to be expected.

**Comment by morphatic** [comment]
You can resolve this issue by refactoring your store (see potential solution below)

tl;dr

You're right, I don't think this is documented, but under the hood it's using [`lodash.merge`](https://lodash.com/docs/4.17.11#merge) in both the [sync](https://github.com/championswimmer/vuex-persist/blob/master/src/index.ts#L236) and [async](https://github.com/championswimmer/vuex-persist/blob/master/src/index.ts#L169) modes. The merge strategy looks like this:

```js
store.replaceState(merge(store.state, savedState || {}))
```

`savedState` represents whatever state was retrieved from storage. As the `lodash` docs explain:

> This method is like `_.assign` except that it recursively merges own and inherited enumerable string keyed properties of source objects into the destination object. Source properties that resolve to undefined are skipped if a destination value exists. Array and plain object properties are merged recursively. Other objects and value types are overridden by assignment.

Can't say for sure if this is the real reason, but given that this plugin allows you to persist different `vuex` modules to different storage locations, my guess is that merge was implemented rather than replace because otherwise, you might end up overwriting the whole state with only a fragment from one of those locations.

Note that this is only what happens when `strictMode` is set to `false` (the default, and the correct setting for production environments). When `strictMode` is set to `true` (e.g. during development), the state is completely replaced via the `RESTORE_MUTATION` [as described here](https://github.com/championswimmer/vuex-persist#support-strict-mode).

If you still want to use this plugin, here's an idea for refactoring your state that should address your issue. Instead of storing the elements in `left` and `right` arrays as you did, why not store them like this:

```js
const state = {
  a: 'left',
  b: 'left',
  c: 'right'
}
```

Then you could write getters for your module that converts them into arrays like you have, e.g.:

```js
const getters = {
  leftSide: state => Object.keys(state).filter(k => state[k] === 'left'),
  rightSide: state => Object.keys(state).filter(k => state[k] === 'right')
}
```

HTH

**Comment by tmladek** [comment]
Cannot confirm that this only happens with `strictMode` set to false, I encountered this in development as well, where I have both VueX strict mode on and consequently `vuex-persist`'s strict mode on.

It is entirely possible to work around this, but it is behavior that caught me off guard and ideally should be at least prominently documented.
